cpp/src/arrow/dataset/dataset.h

```diff
@@ -81,9 +81,32 @@ class ARROW_DS_EXPORT DataSource {
   virtual std::unique_ptr<DataFragmentIterator> GetFragments(
       std::shared_ptr<ScanOptions> options) = 0;
 
+  /// \brief An expression which evaluates to true for all data viewed by this DataSource.
+  /// May be null, which indicates no information is available.
+  const std::shared_ptr<Expression>& partition_expression() const {
+    return partition_expression_;
+  }
+
+  /// FIXME(bkietz) providing a simple mutator like this is probably not ideal
```

Comment:
I'd say part of the parent constructor. I wonder if there's cases where we'd know at runtime.

Reply:
You might acquire statistics at runtime which allow you to refine this expression, for example